libdnf/base/solver_problem.cpp

```diff
@@ -169,5 +264,98 @@ std::string SolverProblem::to_string() const {
     return output;
 }
 
+void SolverProblem::Impl::set_solver_problems(const libdnf::BaseWeakPtr & base, rpm::solv::GoalPrivate & solved_goal) {
```

Comment:
Related to my comment about whether the p_impl is needed, this method is the reason for it.

The `SolverProblems` class is conceptually a data holder and from that perspective it should be constructed and filled up from a business logic class (in this case most likely `GoalPrivate`), it shouldn't have itself a method that takes the business logic class as an argument and then calls its methods. This leads to all sorts of inter-class dependencies going in various directions, insted of `GoalPrivate` having its place in the class hierarchy and all "dependency arrows" pointing in singe direction.

In this case it would IMO mean moving this `set_solver_problems()` method to `GoalPrivate` and the `SolverProblems` class having a method along the lines of `add_problem()` that the `GoalPrivate` class would use to fill it up.

Reply:
This is not an easy decision. I have a plan to use GoalPrivat also for modules. But it requires to keep any rpm specific code that will be in conflict with modules outside of GoalPrivate class.